clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -141,10 +141,8 @@ public void ensureFreshMetadata() {
      * on the current poll if one is active, or the next poll.
      */
     public void wakeup() {
-        if (wakeupsEnabled) {
-            this.wakeup.set(true);
-            this.client.wakeup();
-        }
+        this.wakeup.set(true);
+        this.client.wakeup();
```

Comment:
I am with you that ```selectNow``` will clear the previous wakeup call, as stated "Invoking this method clears the effect of any previous invocations of the wakeup method."

What I thought is, assuming ```selectNow``` and ```wakeup``` are atomic (which they should since they are not blocking). Then we could have the following traces:

1)
```
this.wakeup.set(true);    // in wakeup()
this.client.wakeup();       // in wakeup()
if (wakeup.get())...          // in poll()
this.client.poll(0);   // in poll()
```

and hence it will throw an exception in step 3) without going to step 4). Or in another case:

2)
```
if (wakeup.get())...          // in poll()
this.wakeup.set(true);    // in wakeup()
this.client.poll(0);   // in poll()
this.client.wakeup();       // in wakeup()
```

Then the wakeup call in step 4) will be maintained for the next poll. And:

3)
```
if (wakeup.get())...          // in poll()
this.wakeup.set(true);    // in wakeup()
this.client.wakeup();       // in wakeup()
this.client.poll(0);   // in poll()
```

This is the problem, and even delaying the wakeup will not help here.